### ISSUE TITLE
Views created by mapping over reactive, signal-"controlled" range behave incorrectly

### ISSUE THREAD

**Issue by feathecutie** [leptos-rs/leptos#1789]
**Describe the bug**
When creating views by mapping over a reactive range like `0..some_reactive_value()` or similar and then increasing `some_reactive_value`, only the "newest" view is correctly reactive (e.g. toggles classes/styles when using the `class:` or `style:` syntax), while all "previous" views only seem partially reactive (e.g. don't toggle classes/styles when using the `class:` or `style:` syntax, even though the respective signal was changed).

**Leptos Dependencies**

```toml
leptos = { version = "0.5.0-rc3", features = ["csr", "nightly"] }
```

**To Reproduce**
It is hard to accurately describe steps to reproduce this, but [this gist showcases the simplest version of this bug occuring](https://gist.github.com/feathecutie/0834ffc93c6837d5099aafbb5b175b93).
This gist represents the `main.rs` in a project that only uses Leptos CSR.
I tried using a code sandbox instead, but I couldn't get it to behave how I wanted it to.

**Expected behavior**
All views/components created should continue behaving as intended, e.g. still respecting classes and styles that are to be set reactively via `class:` or `style:`.  

**Screenshots**
See the gist mentioned above for sample code.

**Additional context**
The context I first experienced this bug in was far more complex, so while the example I provided might seem like nobody would ever intentionally design code like this, there certainly are situations where this bug can really get in the way. If wanted, I can provide by concrete example as well.

**Comment by gbj** [comment]
This example can be written in a much simpler way that works and is more correct:
```rust
#[component]
pub fn App() -> impl IntoView {
    let (bg, set_bg) = create_signal(false);
    let (num, set_num) = create_signal(3);

    view! {
        <For
            each=move || 0..num()
            key=|button| *button
            view=move |_| {
                view! {
                    <span
                        style:background-color=move || if bg() {"red"} else {"white"}
                    >
                        Test
                    </span>
                }
            }
        />
        <br/>
        <button on:click=move |_| set_bg.update(|bg| *bg = !*bg)>Toggle style signal</button>
        <br/>
        <button on:click=move |_| set_num.update(|num| *num += 1)>Add new test span</button>
    }
}

fn main() {
    mount_to_body(App)
}
```
Another correct implementation uses plain iteration without the For component:
```rust
#[component]
pub fn App() -> impl IntoView {
    let (bg, set_bg) = create_signal(false);
    let (num, set_num) = create_signal(3);

    let buttons = move || {
        (0..num())
            .map(|i| {
                (
                    i,
                    view! {
                        <span
                            style:background-color=move || if bg() {"red"} else {"white"}
                        >Test </span>
                    },
                )
            })
            .collect::<Vec<_>>()
    };

    view! {
        {buttons}
        <br/>
        <button on:click=move |_| set_bg.update(|bg| *bg = !*bg)>Toggle style signal</button>
        <br/>
        <button on:click=move |_| set_num.update(|num| *num += 1)>Add new test span</button>
    }
}
```
I believe the issue in the provided gist is that the effects to update the styles are created in the scope of the effect that ends up tracking `buttons` inside the `For` component, not in the scope of the rows themselves. This means that they are disposed on each rerun of `buttons`, but because they're then passed through a keyed `For`, the rows that didn't change aren't actually re-run, so you don't re-create the effects.

Basically there are two ways to iterate
1) collecting an iterator of views into a Vec
2) passing data into For and rendering a view per row

This example behaves as expected/in the way that the reactive system is designed, but is broken because it mixes these two paradigms.

**Comment by feathecutie** [comment]
Thanks for the detailed response, this was really helpful!

My usecase unfortunately still required a bit of tinkering, as I essentially had a Signal of type `Vec` that i mapped into views (let's call them nodes), but I need to have another view (let's call it button) displayed inbetween each of these nodes, so up until now I simply created n-1 additional views (buttons) that i interleaved with the `Vec` of views (nodes) I just created.

I figured this would still be fine with `For`, which I really want to use as I don't want my entire list to be rerendered when a new node is added, but I guess i overread the line

> view: renders each T into a view

in the docs of `For`.

I might try to create a PR for the docs, as it wasn't immeatiately obvious for me the "location" where I create my views makes such a big difference, but this might also already be documented somewhere, I don't know.

Btw, my new solution is to create a vector of data, which still contains interspersed nodes and buttons, but to only create the views inside the `For` and it works perfectly!
I solved the issue of the data for my nodes and the data for my buttons having different types (which was the main reason I used views to begin with, so I can store both in the same vec) by simply creating an enum wrapping both types and storing that in the vec. 

I'll close this issue then, since this is indeed intended behaviour and not a bug.
